src/client/devapi_base.cpp

```diff
@@ -4480,234 +5146,271 @@ namespace Tango {
             }
-            catch (CORBA::COMM_FAILURE &comm) {
-                if (comm.minor() == omni::COMM_FAILURE_WaitingForReply) {
-                    TRANSIENT_NOT_EXIST_EXCEPT(comm, "DeviceProxy", "write_pipe", this);
-                } else {
-                    dev_pipe.get_root_blob().reset_insert_ctr();
-                    delete tmp_ptr;
-
-                    set_connection_state(CONNECTION_NOTOK);
-                    stringstream desc;
-                    desc << "Failed to write_pipe on device " << device_name;
-                    ApiCommExcept::re_throw_exception(comm, "API_CommunicationFailed", desc.str(),
-                                                      "DeviceProxy::write_pipe()");
-                }
+        }
+        catch (CORBA::COMM_FAILURE &comm)
+        {
+            if (comm.minor() == omni::COMM_FAILURE_WaitingForReply)
+            {
+                TRANSIENT_NOT_EXIST_EXCEPT(comm, "DeviceProxy", "write_pipe", this);
             }
-            catch (CORBA::SystemException &ce) {
+            else
+            {
                 dev_pipe.get_root_blob().reset_insert_ctr();
-                delete tmp_ptr;
 
                 set_connection_state(CONNECTION_NOTOK);
                 stringstream desc;
                 desc << "Failed to write_pipe on device " << device_name;
-                ApiCommExcept::re_throw_exception(ce, "API_CommunicationFailed", desc.str(),
+                ApiCommExcept::re_throw_exception(comm, "API_CommunicationFailed", desc.str(),
                                                   "DeviceProxy::write_pipe()");
             }
         }
+        catch (CORBA::SystemException &ce)
+        {
+            dev_pipe.get_root_blob().reset_insert_ctr();
+
+            set_connection_state(CONNECTION_NOTOK);
+            stringstream desc;
+            desc << "Failed to write_pipe on device " << device_name;
+            ApiCommExcept::re_throw_exception(ce, "API_CommunicationFailed", desc.str(),
+                                              "DeviceProxy::write_pipe()");
+        }
+    }
+
+    dev_pipe.get_root_blob().reset_insert_ctr();
+
+}
+
+DevPipeData
+DeviceProxy::createDevPipeData(DevicePipe &dev_pipe) const
+{
+    DevPipeData result;
 
-        dev_pipe.get_root_blob().reset_insert_ctr();
-        delete tmp_ptr;
+    result.name = dev_pipe.get_name().c_str();
+    const string &bl_name = dev_pipe.get_root_blob().get_name();
+    if (bl_name.size() != 0)
+        result.data_blob.name = bl_name.c_str();
+
+
+    DevVarPipeDataEltArray *tmp_ptr = dev_pipe.get_root_blob().get_insert_data();
+    if (tmp_ptr == Tango_nullptr)
+    {
+        Except::throw_exception(API_PipeNoDataElement, "No data in pipe!", "DeviceProxy::write_pipe()");
     }
 
+    DevULong max, len;
+    max = tmp_ptr->maximum();
+    len = tmp_ptr->length();
+    result.data_blob.blob_data.replace(max, len, tmp_ptr->get_buffer((DevBoolean) true), true);
+    delete tmp_ptr;
```

Comment:
Actually it seems to be fine to delete _tmp_ptr_ just after _tmp_ptr->get_buffer(true)_ in this case.
The parameter in _get_buffer()_ is named _orphan_. If _orphan_ is true, _tmp_ptr_ is no longer responsible for the memory which was previously allocated for its buffer sequence. _tmp_ptr->get_buffer(true)_ returns a pointer to the data which was previously allocated for _tmp_ptr_ buffer sequence and resets _tmp_ptr_ sequence buffer (pointer set to 0 and sequence length reset to 0 without deallocating the previously allocated memory).
_result.data_blob.blob_data_ is now responsible for deallocating the memory which was previously belonging to _tmp_ptr_.
The last parameter from replace method (_release_ parameter) is important here too. When set to true as here, it indicates that the sequence result.data_blob.blob_data will take care of deallocating the memory of its buffer.
It looks like the previous code was written that way (deleting tmp_ptr later) just to be on the safe side. We can wait for @taurel's review to be sure.